File: agentproto/frame.go
package agentproto

import (
	"encoding/binary"
	"fmt"
)

// Opcode is the first byte of a binary WS frame on the PTY stream (§4.2). The
// opcode discriminates the payload so the hot path carries raw PTY bytes without
// the base64/JSON bloat a text frame would impose.
type Opcode byte

const (
	// OpPTYOut (server → client) prefixes verbatim PTY output bytes. The client
	// feeds them to its terminal emulator (ui/termpane, xterm.js); the server
	// never renders a grid.
	OpPTYOut Opcode = 0x00
	// OpInput (client → server) prefixes raw key bytes. Multi-writer: the server
	// accepts input from ANY connected client — there is no lease gating (§4.2,
	// superseded by Sachin's multi-writer decision).
	OpInput Opcode = 0x01
	// OpResize (client → server) carries a rows,cols uint16 pair (see
	// resizePayloadLen) from any connected client; the server applies
	// last-resize-wins and echoes the authoritative size back as a ResizeMessage.
	OpResize Opcode = 0x02
	// OpRepaint (server → client) prefixes a one-shot screen repaint (clear + the
	// current pane content) sent to a FRESH subscriber before any live output —
	// pipe-pane carries no history, so without it a just-opened pane renders blank
	// (#1592 Phase 2 PR6). The client feeds it to the emulator exactly like OpPTYOut
	// but must NOT count it toward its replay cursor: a repaint is per-subscriber and
	// is not part of the shared ring's monotonic seq, so counting it would desync the
	// ?since arithmetic.
	OpRepaint Opcode = 0x03
	// OpHello (server → client) is the FIRST frame on every new subscription: an
	// 8-byte big-endian uint64 carrying the subscription's starting seq. It delivers
	// in-band the same value the X-Af-Stream-Seq handshake header carries, because a
	// browser WebSocket cannot read 101-handshake response headers and so cannot
	// otherwise learn its absolute cursor to seed ?since replay (#1592 Phase 5 PR1,
	// design §4.3). It carries no PTY bytes and is NOT counted toward the replay
	// cursor. Strictly additive: appended after OpRepaint, existing ops unchanged.
	// Stream consumers that don't handle it (the TUI/apiclient path) ignore it — the
	// opcode decodes cleanly, so ReadMessage never errors and their frame switch
	// simply skips it (unknown-op tolerance without any client change).
	OpHello Opcode = 0x04
)

// resizePayloadLen is the fixed body size of an OpResize frame: two big-endian
// uint16s, rows then cols.
const resizePayloadLen = 4

// helloPayloadLen is the fixed body size of an OpHello frame: one big-endian
// uint64, the subscription's starting seq.
const helloPayloadLen = 8

// String renders an opcode for diagnostics.
func (o Opcode) String() string {
	switch o {
	case OpPTYOut:
		return "PTY_OUT"
	case OpInput:
		return "INPUT"
	case OpResize:
		return "RESIZE"
	case OpRepaint:
		return "REPAINT"
	case OpHello:
		return "HELLO"
	default:
		return fmt.Sprintf("Opcode(0x%02x)", byte(o))
	}
}

// Frame is a decoded binary PTY frame. For OpPTYOut and OpInput the payload is in
// Data; for OpResize the size is in Rows/Cols and Data is nil. Build frames with
// PTYOutFrame/InputFrame/ResizeFrame and serialize with Encode; DecodeFrame is the
// inverse, so DecodeFrame(f.Encode()) round-trips f.
type Frame struct {
	Op   Opcode
	Data []byte // OpPTYOut, OpInput, OpRepaint: raw payload. OpResize/OpHello: nil.
	Rows uint16 // OpResize only.
	Cols uint16 // OpResize only.
	Seq  uint64 // OpHello only.
}

// PTYOutFrame wraps verbatim PTY output bytes (server → client).
func PTYOutFrame(b []byte) Frame { return Frame{Op: OpPTYOut, Data: b} }

// HelloFrame wraps a subscription's starting seq (server → client), emitted as the
// first frame so a client that cannot read the X-Af-Stream-Seq handshake header (a
// browser WebSocket) still learns its absolute cursor to seed ?since replay.
func HelloFrame(seq uint64) Frame { return Frame{Op: OpHello, Seq: seq} }

// RepaintFrame wraps a one-shot screen repaint (server → client): rendered like
// PTY_OUT but NOT counted toward the client's replay cursor (§ OpRepaint).
func RepaintFrame(b []byte) Frame { return Frame{Op: OpRepaint, Data: b} }

// InputFrame wraps raw key bytes (client → server, accepted from any client).
func InputFrame(b []byte) Frame { return Frame{Op: OpInput, Data: b} }

// ResizeFrame wraps a terminal size (client → server, accepted from any client;
// last-resize-wins).
func ResizeFrame(rows, cols uint16) Frame {
	return Frame{Op: OpResize, Rows: rows, Cols: cols}
}

// Encode serializes the frame to its opcode-prefixed wire form.
func (f Frame) Encode() []byte {
	switch f.Op {
	case OpResize:
		out := make([]byte, 1+resizePayloadLen)
		out[0] = byte(OpResize)
		binary.BigEndian.PutUint16(out[1:3], f.Rows)
		binary.BigEndian.PutUint16(out[3:5], f.Cols)
		return out
	case OpHello:
		out := make([]byte, 1+helloPayloadLen)
		out[0] = byte(OpHello)
		binary.BigEndian.PutUint64(out[1:9], f.Seq)
		return out
	default:
		out := make([]byte, 1+len(f.Data))
		out[0] = byte(f.Op)
		copy(out[1:], f.Data)
		return out
	}
}

// DecodeFrame parses an opcode-prefixed binary frame. It errors on an empty frame,
// an unknown opcode, or a malformed OpResize/OpHello body.
func DecodeFrame(raw []byte) (Frame, error) {
	if len(raw) == 0 {
		return Frame{}, fmt.Errorf("agentproto: empty binary frame")
	}
	op := Opcode(raw[0])
	body := raw[1:]
	switch op {
	case OpPTYOut, OpInput, OpRepaint:
		return Frame{Op: op, Data: body}, nil
	case OpResize:
		if len(body) != resizePayloadLen {
			return Frame{}, fmt.Errorf("agentproto: RESIZE frame body is %d bytes, want %d", len(body), resizePayloadLen)
		}
		return Frame{
			Op:   OpResize,
			Rows: binary.BigEndian.Uint16(body[0:2]),
			Cols: binary.BigEndian.Uint16(body[2:4]),
		}, nil
	case OpHello:
		if len(body) != helloPayloadLen {
			return Frame{}, fmt.Errorf("agentproto: HELLO frame body is %d bytes, want %d", len(body), helloPayloadLen)
		}
		return Frame{Op: OpHello, Seq: binary.BigEndian.Uint64(body[0:8])}, nil
	default:
		return Frame{}, fmt.Errorf("agentproto: unknown opcode 0x%02x", byte(op))
	}
}

File: web/src/frame.ts
// A faithful, byte-for-byte port of the Go PTY frame codec (agentproto/frame.go).
// The browser web client (#1592 Phase 5) is a second thin client of the same
// WS PTY protocol the TUI/apiclient speak, so it must encode/decode frames
// identically to the daemon. This is the one place the browser reimplements Go, so
// it is pinned to the Go source of truth by a golden-vector test (frame.test.ts)
// that validates both codecs against the SAME fixture (agentproto/testdata/
// frame_vectors.json). Keep this in lockstep with frame.go; do NOT renumber ops.

/** Opcode is the first byte of a binary WS frame on the PTY stream. */
export enum Op {
  /** server → client: verbatim PTY output bytes. */
  PTYOut = 0x00,
  /** client → server: raw key bytes (multi-writer, accepted from any client). */
  Input = 0x01,
  /** client → server: a rows,cols uint16 pair (last-resize-wins). */
  Resize = 0x02,
  /** server → client: one-shot fresh-subscriber repaint; rendered but NOT counted
   *  toward the replay cursor. */
  Repaint = 0x03,
  /** server → client: the subscription's starting seq as a big-endian uint64, sent
   *  as the FIRST frame so a browser — which cannot read the X-Af-Stream-Seq
   *  handshake header — learns its absolute cursor to seed `?since` replay
   *  (#1592 Phase 5 PR1, design §4.3). NOT counted toward the replay cursor. */
  Hello = 0x04,
}

/** The fixed body size of a RESIZE frame: two big-endian uint16s (rows, cols). */
export const RESIZE_PAYLOAD_LEN = 4;
/** The fixed body size of a HELLO frame: one big-endian uint64 (start seq). */
export const HELLO_PAYLOAD_LEN = 8;

/**
 * A decoded binary PTY frame. Mirrors Go's agentproto.Frame: `data` carries the
 * payload for PTYOut/Input/Repaint; `rows`/`cols` carry a Resize; `seq` carries a
 * Hello. Unused fields hold their zero values (empty data, 0, 0n) so a decoded
 * frame compares cleanly against a constructed one.
 */
export interface Frame {
  op: Op;
  data: Uint8Array;
  rows: number;
  cols: number;
  seq: bigint;
}

const EMPTY = new Uint8Array(0);

function makeFrame(f: { op: Op; data?: Uint8Array; rows?: number; cols?: number; seq?: bigint }): Frame {
  return {
    op: f.op,
    data: f.data ?? EMPTY,
    rows: f.rows ?? 0,
    cols: f.cols ?? 0,
    seq: f.seq ?? 0n,
  };
}

/** Wraps verbatim PTY output bytes (server → client). */
export function ptyOutFrame(data: Uint8Array): Frame {
  return makeFrame({ op: Op.PTYOut, data });
}

/** Wraps raw key bytes (client → server). */
export function inputFrame(data: Uint8Array): Frame {
  return makeFrame({ op: Op.Input, data });
}

/** Wraps a one-shot screen repaint (server → client); rendered like PTYOut but NOT
 *  counted toward the replay cursor. */
export function repaintFrame(data: Uint8Array): Frame {
  return makeFrame({ op: Op.Repaint, data });
}

/** Wraps a terminal size (client → server; last-resize-wins). */
export function resizeFrame(rows: number, cols: number): Frame {
  return makeFrame({ op: Op.Resize, rows, cols });
}

/** Wraps a subscription's starting seq (server → client), the in-band cursor seed. */
export function helloFrame(seq: bigint): Frame {
  return makeFrame({ op: Op.Hello, seq });
}

/** Serializes a frame to its opcode-prefixed wire form (byte-identical to Go's
 *  Frame.Encode). */
export function encode(f: Frame): Uint8Array {
  switch (f.op) {
    case Op.Resize: {
      const out = new Uint8Array(1 + RESIZE_PAYLOAD_LEN);
      out[0] = Op.Resize;
      const dv = new DataView(out.buffer);
      dv.setUint16(1, f.rows, false); // big-endian
      dv.setUint16(3, f.cols, false);
      return out;
    }
    case Op.Hello: {
      const out = new Uint8Array(1 + HELLO_PAYLOAD_LEN);
      out[0] = Op.Hello;
      const dv = new DataView(out.buffer);
      dv.setBigUint64(1, f.seq, false); // big-endian
      return out;
    }
    default: {
      // PTYOut / Input / Repaint (and any op): opcode byte then the raw payload.
      const out = new Uint8Array(1 + f.data.length);
      out[0] = f.op;
      out.set(f.data, 1);
      return out;
    }
  }
}

/**
 * Parses an opcode-prefixed binary frame (the inverse of encode; mirrors Go's
 * DecodeFrame including its error behavior). Throws on an empty frame, an unknown
 * opcode, or a malformed RESIZE/HELLO body.
 */
export function decode(raw: Uint8Array): Frame {
  if (raw.length === 0) {
    throw new Error("agentproto: empty binary frame");
  }
  const op = raw[0] as Op;
  const body = raw.subarray(1);
  switch (op) {
    case Op.PTYOut:
    case Op.Input:
    case Op.Repaint:
      // Copy so the returned frame does not alias the caller's buffer.
      return makeFrame({ op, data: body.slice() });
    case Op.Resize: {
      if (body.length !== RESIZE_PAYLOAD_LEN) {
        throw new Error(`agentproto: RESIZE frame body is ${body.length} bytes, want ${RESIZE_PAYLOAD_LEN}`);
      }
      const dv = new DataView(body.buffer, body.byteOffset, body.byteLength);
      return makeFrame({ op, rows: dv.getUint16(0, false), cols: dv.getUint16(2, false) });
    }
    case Op.Hello: {
      if (body.length !== HELLO_PAYLOAD_LEN) {
        throw new Error(`agentproto: HELLO frame body is ${body.length} bytes, want ${HELLO_PAYLOAD_LEN}`);
      }
      const dv = new DataView(body.buffer, body.byteOffset, body.byteLength);
      return makeFrame({ op, seq: dv.getBigUint64(0, false) });
    }
    default:
      throw new Error(`agentproto: unknown opcode 0x${(op as number).toString(16).padStart(2, "0")}`);
  }
}

/** Renders an opcode for diagnostics (mirrors Go's Opcode.String). */
export function opName(op: Op): string {
  switch (op) {
    case Op.PTYOut:
      return "PTY_OUT";
    case Op.Input:
      return "INPUT";
    case Op.Resize:
      return "RESIZE";
    case Op.Repaint:
      return "REPAINT";
    case Op.Hello:
      return "HELLO";
    default:
      return `Opcode(0x${(op as number).toString(16).padStart(2, "0")})`;
  }
}
